### ISSUE TITLE
proxyRequest with Route headers

### ISSUE THREAD

**Issue by jfoy** [davehorton/drachtio-srf#77]
I'm setting up a basic SIP INVITE dialog between two test components and Drachtio using `proxyRequest`. Eventually I'll want to mutate SIP headers, but for now I'm just establishing a basic flow.

My intent is to deliver an INVITE `cscf-a` -> Drachtio -> `cscf-b`. I'm starting with `cscf-a` at 172.17.0.9 sending the following test message to Drachtio:

```
INVITE sip:4253334455;phone-context=fake.net@fake.net SIP/2.0
Via: SIP/2.0/TCP 172.17.0.9:5060;branch=z9hG4bKisNgfWlI0iyhB-RafsrnDA
From: <sip:+12065551000@fake.net>;tag=sometag
To: <sip:4253334455;phone-context=fake.net@fake.net>
Call-ID: somecallid
CSeq: 1 INVITE
Max-Forwards: 70
Contact: <sip:cscf-a:5060;transport=TCP>
Route: <sip:tas:5060;lr;transport=TCP>, <sip:cscf-b:5060;lr;transport=TCP>
```

`tas` resolves to virtual IP 10.96.233.177 in front of Drachtio, which is on 172.17.0.6. Here's the relevant section from `drachtio.conf.xml`:
```
        <contacts>
            <contact external-ip="10.96.233.177" dns-names="tas">sip:*</contact>
        </contacts>
```

And here's all of `server.js`:
```
const Srf = require('drachtio-srf');
const srf = new Srf();

srf.connect({ host: '127.0.0.1', port: 9021, secret: 'cymru' });

srf
  .on('connect', (err, hostport) => { console.log(`connected: ${hostport}`); })
  .on('error', (err) => { console.log(`Error connecting to drachtio: ${err}`); });

srf.invite((req, res) => {
  console.log("INVITE:", JSON.stringify(req));
  srf.proxyRequest(req).then((result) => console.log("Result:", JSON.stringify(result)) );
});
```

I expected Drachtio to forward the message to `cscf-b` as the next element in the Route, but instead it's sending the below directly to `fake.net`. 
```
INVITE sip:4253334455;phone-context=fake.net@fake.net SIP/2.0
Via: SIP/2.0/UDP 10.96.233.177;rport;branch=z9hG4bK91b58c4e-a71e-475f-913f-33e859e7c1a6
Via: SIP/2.0/TCP 172.17.0.9:5060;branch=z9hG4bKisNgfWlI0iyhB-RafsrnDA;rport=34946
Route: <sip:cscf-b:5060;lr;transport=TCP>
Max-Forwards: 69
From: <sip:+12065551000@fake.net>;tag=sometag
To: <sip:4253334455;phone-context=fake.net@fake.net>
Call-ID: somecallid
CSeq: 1 INVITE
Contact: <sip:cscf-a:5060;transport=TCP>
Content-Length: 0
```

 Am I understanding the intended behavior of `proxyRequest` correctly?

**Comment by davehorton** [comment]
drachtio does not have any logic to handle unsolicited Route headers on a new INVITE, which it seems your scenario is dependent on.  

(not related to this, the drachtio server config works differently than you think.  When you put dns-names="tas" in the contact, you are telling drachtio that "tas" is a dns name that resolves to this local sip server.  And when you put "external-ip" you are telling drachtio to use that address in its Via et).

I've been wondering -- is there an RFC that describes the the accepted use of unsolicited Route headers?  I'd be happy to add support if there is RFC guidance on these scenarios

**Comment by jfoy** [comment]
Thanks for the guidance. I'll have to do some research on what the RFCs specify; I'm new to the domain.

**Comment by jfoy** [comment]
Thanks for the guidance. I'll have to do some research on what the RFCs specify; I'm new to the domain.

**Comment by jfoy** [comment]
I'm not sure what you mean by unsolicited -- do you mean outside of the context of an established dialog? [RFC 3261](https://tools.ietf.org/html/rfc3261) discusses the properties of "loose routers". As I read it, initial Route headers of an INVITE may describe policy for the set and order of proxies that must be visited before a message is delivered to the ultimate recipient in Request-URI.

**Comment by davehorton** [comment]
Route headers are generally calculated based on the Record-Route headers of response to INVITEs, dictating the path of messages in an established dialog.

On Apr 29, 2020, at 7:15 PM, Jack Foy <notifications@github.com> wrote:


I'm not sure what you mean by unsolicited -- do you mean outside of the context of an established dialog? RFC 3261 <https://tools.ietf.org/html/rfc3261> discusses the properties of "loose routers". As I read it, initial Route headers of an INVITE may describe policy for the set and order of proxies that must be visited before a message is delivered to the ultimate recipient in Request-URI.

**Comment by jfoy** [comment]
Yeah, that makes sense for construction of a return path. It looks to me like section 12.2.1.1 describes a use case for Route: headers during initial dialog construction:

>   If the route set is not empty, and the first URI in the route set
>    contains the lr parameter (see Section 19.1.1), the UAC MUST place
>    the remote target URI into the Request-URI and MUST include a Route
>    header field containing the route set values in order, including all
>    parameters.

And in section 8.1.2:

>    If the request
>    contains a Route header field, the request SHOULD be sent to the
>    locations derived from its topmost value, but MAY be sent to any
>    server that the UA is certain will honor the Route and Request-URI
>    policies specified in this document (as opposed to those in RFC
>    2543).  In particular, a UAC configured with an outbound proxy SHOULD
>    attempt to send the request to the location indicated in the first
>    Route header field value instead of adopting the policy of sending
>    all messages to the outbound proxy.
> 
>       This ensures that outbound proxies that do not add Record-Route
>       header field values will drop out of the path of subsequent
>       requests.  It allows endpoints that cannot resolve the first Route
>       URI to delegate that task to an outbound proxy.

If Route: isn't the preferred mechanism, how should we specify this kind of policy?

**Comment by davehorton** [comment]
12.2.1.1 is specifically about a UAS responding to an INVITE (not proxying it on).  This section has the rules for constructing the Route set from the Record-Route headers that the UAS will use in sending further requests _within_the dialog back towards the UAC.  This section is therefore really about the primary / intended purpose of Route headers - to route requests within a dialog.

It seems like you are in an IMS network.  Perhaps in that network whatever vendor gear you are using for the cscf will dictate that you have to support the unsolicited Route header.  I can make changes to support this if neeeded, but I prefer to be guided by RFCs whenever possible.

In a "regular" sip network, I would say if you had the desire to "deliver an INVITE cscf-a -> Drachtio -> cscf-b", the drachtio server would essentially be an outbound proxy and the INVITE sent from cscf-a to drachtio would simply look like this:

```
INVITE sip:4253334455;phone-context=fake.net@cscf-b SIP/2.0
Via: SIP/2.0/TCP 172.17.0.9:5060;branch=z9hG4bKisNgfWlI0iyhB-RafsrnDA
From: <sip:+12065551000@fake.net>;tag=sometag
To: <sip:4253334455;phone-context=fake.net@cscf-b>
Call-ID: somecallid
CSeq: 1 INVITE
Max-Forwards: 70
Contact: <sip:cscf-a:5060;transport=TCP>
```

Srf#proxyRequest would then proxy the invite on to cscf-b

**Comment by davehorton** [comment]
Just checking back here.....if this is blocking you, I could probably add support for the handling of unsolicited Route headers in a proxy scenario.  LMK

**Comment by jfoy** [comment]
Yeah, that would be very interesting to us. I can't tell where this support would go -- what do you think?
